### PR TITLE
Basic Output System + printf/printk

### PR DESCRIPTION
**Pull request by kjhayes**
Very basic "early_output" system for drivers to register "putchar" functions with and get output during early boot.

This system lets the "early_putchar" function be an "output agnostic" building block for functions like "early_puts" and "printk"

Also includes a basic formatting for printk and printf with "%c", "%s", and "%p".

### CODE REVIEW

File: include/kernel/early_output.h [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+#include <kernel/error.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+// This is the maximum number of putchar callbacks which can be registered
+#define EARLY_OUTPUT_MAX_PUTCHAR 8
+
+/*
```
Comment (egelja) [inline include/kernel/early_output.h:13]: put this comment at the top of the file.

File: include/kernel/early_output.h [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+kerror_t register_early_putchar(void (*putchar)(char));
+
+/*
+ * Unregisters a "putchar" previously registered by "register_early_putchar"
+ *
+ * Returns:
+ *     KERR_SUCCESS on success
+ *     KERR_NO_EXIST if this callback hasn't been registered
+ */
+kerror_t unregister_early_putchar(void (*putchar)(char));
```
Comment (egelja) [inline include/kernel/early_output.h]:
```suggestion

typedef void (*putchar_t)(char);

/*
 * Registers "putchar" as a callback function for early output
 *
 * Returns:
 *    KERR_SUCCESS on success
 *    KERR_FULL if the maximum number of callbacks have been registered
 *    KERR_EXIST if this callback has already been registered
 */
kerror_t register_early_putchar(putchar_t cb);

/*
 * Unregisters a "putchar" previously registered by "register_early_putchar"
 *
 * Returns:
 *     KERR_SUCCESS on success
 *     KERR_NO_EXIST if this callback hasn't been registered
 */
kerror_t unregister_early_putchar(putchar_t cb);
```

File: include/kernel/error.h [hunk 3]
```diff
@@ -0,0 +1,36 @@
+#ifndef __KERNEL_ERROR_H__
+#define __KERNEL_ERROR_H__
```
Comment (egelja) [inline include/kernel/error.h]:
```suggestion
#ifndef __KERNEL__ERROR_H__
#define __KERNEL__ERROR_H__
```

File: include/kernel/early_output.h [hunk 4]
```diff
@@ -0,0 +1,55 @@
+#ifndef __KERNEL_EARLY_OUTPUT_H__
+#define __KERNEL_EARLY_OUTPUT_H__
```
Comment (egelja) [inline include/kernel/early_output.h]:
```suggestion
#ifndef __KERNEL__EARLY_OUTPUT_H__
#define __KERNEL__EARLY_OUTPUT_H__
```

File: kernel/early_output.c [hunk 5]
```diff
@@ -0,0 +1,56 @@
+
+#include <kernel/early_output.h>
+#include <stddef.h>
+
+static void(*early_putchar_array[EARLY_OUTPUT_MAX_PUTCHAR])(char) = {0};
```
Comment (egelja) [inline kernel/early_output.c]:
```suggestion
static putchar_t early_putchar_array[EARLY_OUTPUT_MAX_PUTCHAR] = {0};
```

File: kernel/early_output.c [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+      return KERR_EXIST;
+    }
+  }
+
+  if(index >= 0) {
+    early_putchar_array[index] = putchar;
+    return KERR_SUCCESS;
+  }
+
+  return KERR_FULL;
```
Comment (egelja) [inline kernel/early_output.c]: I'd invert this
```suggestion
  if(index < 0)
      return KERR_FULL;
      
  early_putchar_array[index] = putchar;
  return KERR_SUCCESS;
```

File: kernel/early_output.c [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+  return KERR_NO_EXIST;
+}
+
+void early_putchar(char c) 
+{
+  for(int i = 0; i < EARLY_OUTPUT_MAX_PUTCHAR; i++) {
+    if(early_putchar_array[i] == NULL) {
+      continue;
+    }
+    (*(early_putchar_array[i]))(c);
```
Comment (egelja) [inline kernel/early_output.c]:
```suggestion
    putchar_t cb = early_putchar_array[i];
    if (cb == NULL)
       continue;

    (*cb)(c);
```

File: libc/include/stdarg.h [hunk 8]
```diff
@@ -6,8 +6,12 @@ extern "C" {
...
 
 #ifdef __GNUC__
-/* We use the GCC stdarg as it is compiler-specific. */
-#    include_next <stdarg.h> /* NOLINT */
+
+typedef __builtin_va_list va_list;
+#define va_start __builtin_va_start
+#define va_end __builtin_va_end
+#define va_arg __builtin_va_arg
+
```
Comment (egelja) [inline libc/include/stdarg.h]: i think you're missing `va_copy` and the definition of `va_start` may be wrong

image

File: libc/src/stdio/printf.c [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+ * Ignoring modifiers for now
+ */
+
+// Prints for "%p" and returns the number of characters printed
+static int pointer_print(void(*puts)(const char*), void *p) 
+{
+  uintptr_t ip = (uintptr_t)p;
+  
+  // 2 hex digits per byte + 2 digits for "0x" + 1 null terminator
+#define BUF_LEN (sizeof(void*) * 2) + 2 + 1
```
Comment (egelja) [inline libc/src/stdio/printf.c]:
```suggestion
  constexpr size_t BUF_LEN = (sizeof(void*) * 2) + 2 + 1;
```

File: libc/src/stdio/printf.c [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+ * %n
+ *
+ * (I don't even want to think about this one)
+ * %f
+ *
+ * Ignoring modifiers for now
+ */
+
+// Prints for "%p" and returns the number of characters printed
+static int pointer_print(void(*puts)(const char*), void *p)
```
Comment (egelja) [inline libc/src/stdio/printf.c]:
```suggestion
typedef void (*puts_t)(const char*);

// Prints for "%p" and returns the number of characters printed
static int pointer_print(puts_t puts, void *p) 
```

File: libc/src/stdio/printf.c [hunk 11]
```diff
@@ -0,0 +1,205 @@
...
+    }
+  }
+
+  buf[0] = '0';
+  buf[1] = 'x';
+  buf[BUF_LEN-1] = '\0';
+  (*puts)(buf);
+ 
+  return BUF_LEN-1;
+#undef BUF_LEN
```
Comment (egelja) [inline libc/src/stdio/printf.c]:
```suggestion
```

File: libc/src/stdio/printf.c [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+}
+
+static int formatted_print(void(*puts)(const char*), const char *fmt, va_list args) 
+{
+#define BUF_LEN 32
```
Comment (egelja) [inline libc/src/stdio/printf.c]:
```suggestion
constexpr size_t BUF_LEN = 32;
```

File: libc/src/stdio/printf.c [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+  buf[0] = '0';
+  buf[1] = 'x';
+  buf[BUF_LEN-1] = '\0';
+  (*puts)(buf);
+ 
+  return BUF_LEN-1;
+#undef BUF_LEN
+}
+
+static int formatted_print(void(*puts)(const char*), const char *fmt, va_list args)
```
Comment (egelja) [inline libc/src/stdio/printf.c]:
```suggestion
static int formatted_print(puts_t puts, const char *fmt, va_list args) 
```

File: libc/src/stdio/printf.c [hunk 14]
```diff
@@ -0,0 +1,205 @@
...
+    buf_head = 0;
+  }
+
+  return num_chars;
+
+#undef BUF_LEN
+}
+
+void printf_puts(const char *s) {
+  // TODO
```
Comment (egelja) [inline libc/src/stdio/printf.c]:
```suggestion
  // TODO
  abort();
```

File: libc/src/stdio/printf.c [hunk 15]
```diff
@@ -0,0 +1,205 @@
...
+  char buf[BUF_LEN+1];
+  int buf_head = 0;
+
+  // Add backup null terminator
+  buf[BUF_LEN] = '\0';
+
+  bool escaped = false;
+  do 
+  {
+    if(escaped) {
```
Comment (egelja) [inline libc/src/stdio/printf.c]:
```suggestion
    if(escaped) { /* we found a % */
```

File: libc/src/stdio/printf.c [hunk 16]
```diff
@@ -0,0 +1,205 @@
...
+
+      // Predefinitions for use in the switch statement
+      char *s;
+      void *p;
+      //
+
+      switch(*fmt) {
+        case '\0':
+          // In this case we need to output the final %
+        case '%':
```
Comment (egelja) [inline libc/src/stdio/printf.c]:
```suggestion
	[[fallthrough]];
        case '%': /* %% */
```

File: libc/src/stdio/printf.c [hunk 17]
```diff
@@ -0,0 +1,205 @@
...
+          buf[0] = '%';
+          buf[1] = '\0';
+          (*puts)(buf);
+          num_chars += 1;
+          escaped = false;
```
Comment (egelja) [inline libc/src/stdio/printf.c]: the `escaped=false` should go after the switch

File: libc/src/stdio/printf.c [hunk 18]
```diff
@@ -0,0 +1,205 @@
...
+          break;
+        case 'c':
```
Comment (egelja) [inline libc/src/stdio/printf.c]:
```suggestion
        case 'c':  /* %c */
```

File: libc/src/stdio/printf.c [hunk 19]
```diff
@@ -0,0 +1,205 @@
...
+          buf[0] = (char)va_arg(args, int);
```
Comment (egelja) [inline libc/src/stdio/printf.c]: are you sure this is an int?

File: libc/src/stdio/printf.c [hunk 20]
```diff
@@ -0,0 +1,205 @@
...
+          buf[1] = '\0';
+          (*puts)(buf);
+          num_chars += 1;
+          escaped = false;
+          break;
+        case 's':
```
Comment (egelja) [inline libc/src/stdio/printf.c]:
```suggestion
        case 's': /* %s */
```

File: libc/src/stdio/printf.c [hunk 21]
```diff
@@ -0,0 +1,205 @@
...
+          buf[1] = *fmt;
+          buf[2] = '\0';
+          num_chars += 2;
+          (*puts)(buf);
+          escaped = false;
+          break;
+      }
+    }
+
+    else if(!escaped && *fmt != '\0') {
```
Comment (egelja) [inline libc/src/stdio/printf.c]:
```suggestion
    else if(*fmt != '\0') {
```

`!escpaed` is true by definition

File: libc/src/stdio/printf.c [hunk 22]
```diff
@@ -0,0 +1,205 @@
...
+          num_chars += 1;
+          escaped = false;
+          break;
+        case 's':
+          s = (char*)va_arg(args,char*);
+          (*puts)(s);
+          num_chars += (int)strlen(s);
+          escaped = false;
+          break;
+        case 'p':
```
Comment (egelja) [inline libc/src/stdio/printf.c]:
```suggestion
        case 'p': /* %p */
```

File: libc/src/stdio/printf.c [hunk 23]
```diff
@@ -0,0 +1,205 @@
...
+          p = (void*)va_arg(args,void*);
+          num_chars += pointer_print(puts,p);
+          escaped = false;
+          break;
+        default:
```
Comment (egelja) [inline libc/src/stdio/printf.c]:
```suggestion
        default: /* unknown specifier */
```

**Review by egelja — DISMISSED** [review]

File: libc/src/stdio/printf.c (hunk 21 above)
Comment (kjhayes) [reply libc/src/stdio/printf.c]: Yeah I used to allow it to flow through in some cases but ended up changing it and forgot to patch that

File: libc/src/stdio/printf.c (hunk 19 above)
Comment (kjhayes) [reply libc/src/stdio/printf.c]: Yeah, va_args promotes all types smaller than `int` to an `int`

File: kernel/early_output.c (hunk 5 above)
Comment (kjhayes) [reply kernel/early_output.c]: I'm not against function pointers without typedefs, but I won't fight you for them

File: libc/src/stdio/printf.c (hunk 22 above)
Comment (kjhayes) [reply libc/src/stdio/printf.c]: This feels a little bit unnecessary but it can't hurt ig

File: libc/src/stdio/printf.c (hunk 17 above)
Comment (kjhayes) [reply libc/src/stdio/printf.c]: No, it just seems like that, remember we're eventually going to have to handle the case of stuff like `%llx` so we'll update state when we see an escaped `l` but not un-escape

File: libc/include/stdarg.h (hunk 8 above)
Comment (kjhayes) [reply libc/include/stdarg.h]: Resolved in a new commit

File: libc/src/stdio/printf.c [hunk 24]
```diff
@@ -0,0 +1,207 @@
...
+  if(buf_head > 0) {
+    buf[buf_head] = '\0';
+    (*puts)(buf);
+    num_chars += buf_head;
+    buf_head = 0;
+  }
+
+  return num_chars;
+
+#undef BUF_LEN
```
Comment (egelja) [inline libc/src/stdio/printf.c]:
```suggestion
```

File: libc/src/stdio/printf.c [hunk 25]
```diff
@@ -0,0 +1,207 @@
...
+#ifdef KERNEL
+
+#include <kernel/early_output.h>
+
+static void printk_puts(const char *s) 
+{
+  early_puts(s);
+}
+
+#endif
```
Comment (egelja) [inline libc/src/stdio/printf.c]:
```suggestion
#endif /* KERNEL */
```

**Review by egelja — DISMISSED** [review]
`make format` please you're failing actions

**Comment by kjhayes** [comment]
Checks are passing now, sorry

**Review by egelja — DISMISSED** [review]
Lgtm

**Review by egelja — APPROVED** [review]
Lgtm

**Pull request merged**